File: Kvasha/1559.js
/**
 * @param {character[][]} grid
 * @return {boolean}
 */
let containsCycle = function (grid) {
  const m = grid.length;
  const n = grid[0].length;
  let visited = new Array(m).fill(0).map((_) => new Array(n));

  const dir = {
    left: { i: 0, j: -1 },
    right: { i: 0, j: 1 },
    up: { i: -1, j: 0 },
    down: { i: 1, j: 0 },
  };

  const dirPrev = {
    left: 'right',
    right: 'left',
    up: 'down',
    down: 'up',
  };

  function dfs(row, col, prev = '') {
    if (visited[row][col]) return true;

    visited[row][col] = true;

    for (let d in dir) {
      if (d === prev) continue;

      let i = row + dir[d].i;
      let j = col + dir[d].j;

      if (i > -1 && i < m && j > -1 && j < n) {
        if (grid[row][col] === grid[i][j] && dfs(i, j, dirPrev[d])) return true;
      }
    }

    visited[row][col] = false;
    return false;
  }

  for (let i = 0; i < m; ++i) {
    for (let j = 0; j < n; ++j) {
      if (visited[i][j] === false) continue;
      if (dfs(i, j)) return true;
    }
  }

  return false;
};

File: Дорофеев/1559.cpp
class Solution
{
public:
    bool answer = false;
    int row, col;

    void DFS(vector<vector<char>>& grid, vector<vector<bool>>& visited,
        int r, int c, int pr, int pc)
    {
        if (visited[r][c])
        {
            answer = true;
            return;
        }
        visited[r][c] = true;
        if (r + 1 < row && r + 1 != pr && grid[r + 1][c] == grid[r][c])
            DFS(grid, visited, r + 1, c, r, c);
        if (c + 1 < col && c + 1 != pc && grid[r][c + 1] == grid[r][c])
            DFS(grid, visited, r, c + 1, r, c);
        if (r - 1 >= 0 && r - 1 != pr && grid[r - 1][c] == grid[r][c])
            DFS(grid, visited, r - 1, c, r, c);
        if (c - 1 >= 0 && c - 1 != pc && grid[r][c - 1] == grid[r][c])
            DFS(grid, visited, r, c - 1, r, c);
    }

    bool containsCycle(vector<vector<char>>& grid)
    {
        row = grid.size();
        col = grid[0].size();

        vector<vector<bool>> visited(row, vector<bool>(col, false));

        for (int i = 0; i < visited.size(); ++i)
            for (int j = 0; j < visited[0].size(); ++j)
                if (!visited[i][j])
                    DFS(grid, visited, i, j, -1, -1);

        return answer;
    }
};